code/game/objects/effects/anomalies.dm

```diff
@@ -370,51 +397,85 @@
-	var/turf/open/T = get_turf(src)
-	var/mob/living/simple_animal/hostile/retaliate/goat/radioactive/S = new(T)
-
-	var/list/mob/dead/observer/candidates = pollCandidatesForMob("Do you want to play as a radioactive goat?", ROLE_SENTIENCE, null, null, 100, S, POLL_IGNORE_PYROSLIME)
-	if(LAZYLEN(candidates))
-		var/mob/dead/observer/chosen = pick(candidates)
-		S.key = chosen.key
-		var/datum/action/cooldown/spell/conjure/radiation_anomaly/spell
-		spell.Grant(S)
-		log_game("[key_name(S.key)] was made into a radioactive goat by radiation anomaly at [AREACOORD(T)].")
+	for(var/i=1 to 15)
+		fire_nuclear_particle()
+		radiation_pulse(src, 20000, 7)
+	if(spawn_goat == ANOMALY_RADIATION_YES_GOAT)
+		var/turf/open/T = get_turf(src)
+		var/mob/living/simple_animal/hostile/retaliate/goat/radioactive/S = new(T)
+
+		var/list/mob/dead/observer/candidates = pollCandidatesForMob("Do you want to play as a radioactive goat?", ROLE_SENTIENCE, null, null, 100, S, POLL_IGNORE_PYROSLIME)
+		if(LAZYLEN(candidates))
+			var/mob/dead/observer/chosen = pick(candidates)
+			S.key = chosen.key
+			var/datum/action/cooldown/spell/conjure/radiation_anomaly/spell
+			spell.Grant(S)
+			log_game("[key_name(S.key)] was made into a radioactive goat by radiation anomaly at [AREACOORD(T)].")
 
 /obj/effect/anomaly/radiation/detonate()
-	INVOKE_ASYNC(src, PROC_REF(rad_Spin))
+	INVOKE_ASYNC(src, PROC_REF(makegoat))
 
-/obj/effect/anomaly/radiation/proc/rad_Spin(increment = 1)
-	if(increment > 100)
-		if(spawn_goat)//only spawn the goat once, when the anomaly explodes
-			INVOKE_ASYNC(src, PROC_REF(makegoat))
-		qdel(src)
-	radiation_pulse(src, 5000, 7)
-	var/turf/T = get_turf(src)
-	var/angle = increment * 10
-	T.fire_nuclear_particle(angle)
-	addtimer(CALLBACK(src, PROC_REF(rad_Spin), increment + 1), 0.7)
-		
+/////////////////////
 
-/obj/effect/anomaly/radiation/process(delta_time)
-	anomalyEffect(delta_time)
-	if(death_time < world.time)
-		if(loc)
-			detonate()
+/obj/effect/anomaly/hallucination
+	name = "hallucination anomaly"
+	icon_state = "hallucination_anomaly"
+	coretype = ANOMALY_HALLUCINATION
+	/// Time passed since the last effect, increased by delta_time of the SSobj
+	var/ticks = 0
+	/// How many seconds between each small hallucination pulses
+	var/release_delay = 5
+
+/obj/effect/anomaly/hallucination/anomalyEffect(delta_time)
+	. = ..()
+	ticks += delta_time
+	if(ticks < release_delay)
+		return
+	ticks -= release_delay
+	var/turf/open/our_turf = get_turf(src)
+	if(istype(our_turf))
+		hallucination_pulse(our_turf, 5)
+
+/obj/effect/anomaly/hallucination/detonate()
+	var/turf/open/our_turf = get_turf(src)
+	if(istype(our_turf))
+		hallucination_pulse(our_turf, 10)
+
+/proc/hallucination_pulse(turf/location, range, strength = 50)
+	for(var/mob/living/carbon/human/near in view(location, range))
+		// If they are immune to hallucinations
+		if (HAS_TRAIT(near, TRAIT_MADNESS_IMMUNE) || (near.mind && HAS_TRAIT(near.mind, TRAIT_MADNESS_IMMUNE)))
```

Comment:
You can just check for TRAIT_MESONS here instead of adding an entirely new trait